plugins/aws/providers.go

```diff
@@ -0,0 +1,142 @@
+}
+
+func NewAssumeRoleProvider(awsConfig *confighelpers.Config, in sdk.ProvisionInput, out *sdk.ProvisionOutput) aws.CredentialsProvider {
+	roleCacheKey := getRoleCacheKey(awsConfig.RoleARN)
+	if in.Cache.Has(roleCacheKey) {
+		return NewStsCacheProvider(roleCacheKey, in.Cache)
+	}
+
+	if awsConfig.HasMfaSerial() && awsConfig.MfaToken != "" {
+		return &assumeRoleProvider{
+			AssumeRoleProvider: confighelpers.AssumeRoleProvider{
+				StsClient:         getSTSClient(awsConfig.Region, NewMFASessionTokenProvider(awsConfig, in, out)),
+				RoleARN:           awsConfig.RoleARN,
+				RoleSessionName:   awsConfig.RoleSessionName,
+				ExternalID:        awsConfig.ExternalID,
+				Duration:          900 * time.Second, // minimum duration of 15 minutes
+				Tags:              awsConfig.SessionTags,
+				TransitiveTagKeys: awsConfig.TransitiveSessionTags,
+				SourceIdentity:    awsConfig.SourceIdentity,
+				Mfa:               &confighelpers.Mfa{},
+			},
+			stsCacheWriter: NewStsCacheWriter(roleCacheKey, out.Cache),
+		}
+	}
+
+	return &assumeRoleProvider{
+		AssumeRoleProvider: confighelpers.AssumeRoleProvider{
+			StsClient:         getSTSClient(awsConfig.Region, NewMasterCredentialsProvider(in.ItemFields)),
+			RoleARN:           awsConfig.RoleARN,
+			RoleSessionName:   awsConfig.RoleSessionName,
+			ExternalID:        awsConfig.ExternalID,
+			Duration:          900 * time.Second, // minimum duration of 15 minutes
+			Tags:              awsConfig.SessionTags,
+			TransitiveTagKeys: awsConfig.TransitiveSessionTags,
+			SourceIdentity:    awsConfig.SourceIdentity,
+			Mfa:               &confighelpers.Mfa{},
+		},
+		stsCacheWriter: NewStsCacheWriter(roleCacheKey, out.Cache),
+	}
+}
+
+type mfaSessionTokenProvider struct {
+	confighelpers.SessionTokenProvider
+	stsCacheWriter
+}
+
+func (p mfaSessionTokenProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	credentials, err := p.SessionTokenProvider.Retrieve(ctx)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	err = p.stsCacheWriter.persist(credentials)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	return credentials, nil
+}
+
+func NewMFASessionTokenProvider(awsConfig *confighelpers.Config, in sdk.ProvisionInput, out *sdk.ProvisionOutput) aws.CredentialsProvider {
+	if in.Cache.Has(mfaCacheKey) {
+		return NewStsCacheProvider(mfaCacheKey, in.Cache)
+	}
+
+	return &mfaSessionTokenProvider{
+		SessionTokenProvider: confighelpers.SessionTokenProvider{
+			StsClient: getSTSClient(awsConfig.Region, NewMasterCredentialsProvider(in.ItemFields)),
+			Duration:  900 * time.Second, // minimum duration of 15 minutes,
+			Mfa:       confighelpers.NewMfa(awsConfig),
+		},
+		stsCacheWriter: NewStsCacheWriter(mfaCacheKey, out.Cache),
+	}
+}
+
+func NewStsCacheProvider(key string, cache sdk.CacheState) aws.CredentialsProvider {
+	return stsCacheProvider{
+		awsCacheKey: key,
+		cache:       cache,
+	}
+}
+
+type masterAwsCredentialsProvider struct {
```

Comment:
We generally try to steer away from using "master" because of its relation to slavery. 

Can we use "AWS credentials" or "Access Keys" (that's what AWS uses for that keypair) instead?

Reply:
Also, there is no such thing as a "master AWS credential" as far as I'm aware.

Reply:
I saw this get called like that in `aws-vault` (https://github.com/99designs/aws-vault/blob/b15cc22022184d7835311e77cd59b720e6007876/vault/vault.go#L243). I'll rename it to `accessKeysProvider`

Reply:
Done